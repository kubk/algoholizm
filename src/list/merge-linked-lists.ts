export class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

const listNodeToArray = (n: ListNode) => {
  const result = [];
  let current: ListNode | null = n;
  do {
    if (current) {
      result.push(current.val);
    }
    current = current.next;
  } while (current);

  return result;
};

const arrayToListNode = (arr: number[]): ListNode | null => {
  if (arr.length === 0) {
    return null;
  }

  let parent: ListNode | null = null;
  let first: ListNode | null = null;

  for (let i = 0; i < arr.length; i++) {
    const n = new ListNode(arr[i]);
    if (parent) {
      parent.next = n;
      parent = n;
    } else {
      parent = n;
      first = n;
    }
  }

  return first;
};

export const mergeTwoLists = (list1: ListNode | null, list2: ListNode | null): ListNode | null => {
  const arr1 = list1 ? listNodeToArray(list1) : [];
  const arr2 = list2 ? listNodeToArray(list2) : [];
  const result = arr1.concat(arr2).sort((a, b) => a - b);

  return arrayToListNode(result);
};
