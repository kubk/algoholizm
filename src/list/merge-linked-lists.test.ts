import { ListNode, mergeTwoLists } from './merge-linked-lists';
import { expect, test } from 'vitest';

test('merge linked lists', () => {
  const l1 = new ListNode(1, new ListNode(2, new ListNode(4)));
  const l2 = new ListNode(1, new ListNode(3, new ListNode(4)));

  expect(mergeTwoLists(l1, l2)).toMatchSnapshot();
});
