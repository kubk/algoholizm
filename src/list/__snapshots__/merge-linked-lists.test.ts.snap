// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`merge linked lists 1`] = `
ListNode {
  "next": ListNode {
    "next": ListNode {
      "next": ListNode {
        "next": ListNode {
          "next": ListNode {
            "next": null,
            "val": 4,
          },
          "val": 4,
        },
        "val": 3,
      },
      "val": 2,
    },
    "val": 1,
  },
  "val": 1,
}
`;
